File: agent_slack/coordination/task.py
"""
Task Primitive

Tasks are units of work that agents can assign to each other.
They provide structure beyond simple messages - tracking state,
ownership, deadlines, and completion.
"""

from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import Optional, Any
from enum import Enum
import uuid


class TaskStatus(Enum):
    """Status of a task."""
    
    PENDING = "pending"          # Created but not started
    ASSIGNED = "assigned"        # Assigned to an agent
    IN_PROGRESS = "in_progress"  # Being worked on
    BLOCKED = "blocked"          # Waiting on something
    IN_REVIEW = "in_review"      # Completed, awaiting review
    COMPLETED = "completed"      # Successfully finished
    CANCELLED = "cancelled"      # Cancelled before completion
    FAILED = "failed"            # Failed to complete


class TaskPriority(Enum):
    """Priority levels for tasks."""
    
    CRITICAL = "critical"    # Drop everything
    HIGH = "high"            # Important, do soon
    MEDIUM = "medium"        # Normal priority
    LOW = "low"              # Do when convenient
    BACKGROUND = "background" # No rush


@dataclass
class TaskComment:
    """A comment or update on a task."""
    
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    agent_id: str = ""
    content: str = ""
    created_at: datetime = field(default_factory=datetime.utcnow)
    
    # Optional: link to message that triggered this comment
    message_id: Optional[str] = None


@dataclass
class Task:
    """
    A task assigned to an agent.
    
    Key concepts:
    - Tasks have clear ownership (assignee)
    - Tasks track state through a lifecycle
    - Tasks can be linked to messages/threads
    - Tasks have structured input/output
    
    Design decisions:
    - Tasks are separate from messages (but can be linked)
    - Multiple agents can be involved (creator, assignee, reviewers)
    - Tasks support blocking/dependencies
    - Progress and output are tracked
    """
    
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    
    # Task description
    title: str = ""
    description: str = ""
    
    # Ownership
    created_by: str = ""             # Agent who created the task
    assigned_to: Optional[str] = None # Agent responsible for completion
    reviewers: list[str] = field(default_factory=list)
    
    # Status and priority
    status: TaskStatus = TaskStatus.PENDING
    priority: TaskPriority = TaskPriority.MEDIUM
    
    # Context links
    channel_id: Optional[str] = None  # Channel where task was created
    thread_id: Optional[str] = None   # Thread for task discussion
    source_message_id: Optional[str] = None  # Message that spawned task
    
    # Timing
    created_at: datetime = field(default_factory=datetime.utcnow)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    due_at: Optional[datetime] = None
    
    # Progress tracking
    progress_percent: int = 0         # 0-100
    status_message: str = ""          # Current status text
    
    # Input/output
    input_data: dict = field(default_factory=dict)
    output_data: dict = field(default_factory=dict)
    
    # Dependencies
    blocked_by: list[str] = field(default_factory=list)  # Task IDs
    blocks: list[str] = field(default_factory=list)      # Task IDs
    
    # History
    comments: list[TaskComment] = field(default_factory=list)
    
    # Metadata
    tags: list[str] = field(default_factory=list)
    metadata: dict = field(default_factory=dict)
    
    @property
    def is_open(self) -> bool:
        """Check if task is still open (not completed/cancelled/failed)."""
        return self.status not in (
            TaskStatus.COMPLETED,
            TaskStatus.CANCELLED,
            TaskStatus.FAILED,
        )
    
    @property
    def is_blocked(self) -> bool:
        """Check if task is blocked."""
        return self.status == TaskStatus.BLOCKED or len(self.blocked_by) > 0
    
    @property
    def is_overdue(self) -> bool:
        """Check if task is past due date."""
        if not self.due_at:
            return False
        return datetime.utcnow() > self.due_at and self.is_open
    
    @property
    def duration(self) -> Optional[timedelta]:
        """Get the duration of the task (if started and completed)."""
        if self.started_at and self.completed_at:
            return self.completed_at - self.started_at
        return None
    
    def assign(self, agent_id: str) -> "Task":
        """Assign the task to an agent."""
        self.assigned_to = agent_id
        self.status = TaskStatus.ASSIGNED
        return self
    
    def start(self) -> "Task":
        """Mark task as started."""
        self.status = TaskStatus.IN_PROGRESS
        self.started_at = datetime.utcnow()
        return self
    
    def update_progress(
        self,
        percent: int,
        message: str = "",
    ) -> "Task":
        """Update task progress."""
        self.progress_percent = min(100, max(0, percent))
        if message:
            self.status_message = message
        return self
    
    def block(self, reason: str, blocked_by_task_id: Optional[str] = None) -> "Task":
        """Mark task as blocked."""
        self.status = TaskStatus.BLOCKED
        self.status_message = reason
        if blocked_by_task_id:
            self.blocked_by.append(blocked_by_task_id)
        return self
    
    def unblock(self, task_id: Optional[str] = None) -> "Task":
        """Unblock the task."""
        if task_id and task_id in self.blocked_by:
            self.blocked_by.remove(task_id)
        
        if not self.blocked_by:
            self.status = TaskStatus.IN_PROGRESS
            self.status_message = ""
        return self
    
    def submit_for_review(self, output: dict = None) -> "Task":
        """Submit task for review."""
        self.status = TaskStatus.IN_REVIEW
        if output:
            self.output_data = output
        self.progress_percent = 100
        return self
    
    def complete(self, output: dict = None) -> "Task":
        """Mark task as completed."""
        self.status = TaskStatus.COMPLETED
        self.completed_at = datetime.utcnow()
        self.progress_percent = 100
        if output:
            self.output_data = output
        return self
    
    def cancel(self, reason: str = "") -> "Task":
        """Cancel the task."""
        self.status = TaskStatus.CANCELLED
        self.completed_at = datetime.utcnow()
        self.status_message = reason
        return self
    
    def fail(self, reason: str = "", error: Any = None) -> "Task":
        """Mark task as failed."""
        self.status = TaskStatus.FAILED
        self.completed_at = datetime.utcnow()
        self.status_message = reason
        if error:
            self.output_data["error"] = str(error)
        return self
    
    def add_comment(
        self,
        agent_id: str,
        content: str,
        message_id: Optional[str] = None,
    ) -> TaskComment:
        """Add a comment to the task."""
        comment = TaskComment(
            agent_id=agent_id,
            content=content,
            message_id=message_id,
        )
        self.comments.append(comment)
        return comment
    
    def to_dict(self) -> dict:
        """Serialize to dictionary."""
        return {
            "id": self.id,
            "title": self.title,
            "description": self.description,
            "created_by": self.created_by,
            "assigned_to": self.assigned_to,
            "reviewers": self.reviewers,
            "status": self.status.value,
            "priority": self.priority.value,
            "channel_id": self.channel_id,
            "thread_id": self.thread_id,
            "progress_percent": self.progress_percent,
            "status_message": self.status_message,
            "is_open": self.is_open,
            "is_blocked": self.is_blocked,
            "is_overdue": self.is_overdue,
            "created_at": self.created_at.isoformat(),
            "started_at": self.started_at.isoformat() if self.started_at else None,
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "due_at": self.due_at.isoformat() if self.due_at else None,
            "tags": self.tags,
            "comment_count": len(self.comments),
        }
    
    def __str__(self) -> str:
        status_icons = {
            TaskStatus.PENDING: "⏸️",
            TaskStatus.ASSIGNED: "📋",
            TaskStatus.IN_PROGRESS: "🔄",
            TaskStatus.BLOCKED: "🚫",
            TaskStatus.IN_REVIEW: "👀",
            TaskStatus.COMPLETED: "✅",
            TaskStatus.CANCELLED: "❌",
            TaskStatus.FAILED: "💥",
        }
        icon = status_icons.get(self.status, "❓")
        assignee = f" -> {self.assigned_to[:8]}" if self.assigned_to else ""
        return f"{icon} [{self.priority.value}] {self.title}{assignee}"


class TaskManager:
    """
    Manages tasks within a workspace.
    
    Handles task lifecycle, assignment, and queries.
    """
    
    def __init__(self):
        self._tasks: dict[str, Task] = {}
        self._by_assignee: dict[str, list[str]] = {}
        self._by_channel: dict[str, list[str]] = {}
    
    def create(
        self,
        title: str,
        created_by: str,
        description: str = "",
        priority: TaskPriority = TaskPriority.MEDIUM,
        assigned_to: Optional[str] = None,
        channel_id: Optional[str] = None,
        **kwargs,
    ) -> Task:
        """Create a new task."""
        task = Task(
            title=title,
            description=description,
            created_by=created_by,
            assigned_to=assigned_to,
            priority=priority,
            channel_id=channel_id,
            **kwargs,
        )
        
        if assigned_to:
            task.status = TaskStatus.ASSIGNED
        
        return self.add(task)
    
    def add(self, task: Task) -> Task:
        """Add a task to the manager."""
        self._tasks[task.id] = task
        
        # Index by assignee
        if task.assigned_to:
            if task.assigned_to not in self._by_assignee:
                self._by_assignee[task.assigned_to] = []
            self._by_assignee[task.assigned_to].append(task.id)
        
        # Index by channel
        if task.channel_id:
            if task.channel_id not in self._by_channel:
                self._by_channel[task.channel_id] = []
            self._by_channel[task.channel_id].append(task.id)
        
        return task
    
    def get(self, task_id: str) -> Optional[Task]:
        """Get a task by ID."""
        return self._tasks.get(task_id)
    
    def get_by_assignee(
        self,
        agent_id: str,
        open_only: bool = True,
    ) -> list[Task]:
        """Get tasks assigned to an agent."""
        task_ids = self._by_assignee.get(agent_id, [])
        tasks = [self._tasks[id] for id in task_ids if id in self._tasks]
        
        if open_only:
            tasks = [t for t in tasks if t.is_open]
        
        return sorted(tasks, key=lambda t: (t.priority.value, t.created_at))
    
    def get_by_channel(
        self,
        channel_id: str,
        open_only: bool = True,
    ) -> list[Task]:
        """Get tasks in a channel."""
        task_ids = self._by_channel.get(channel_id, [])
        tasks = [self._tasks[id] for id in task_ids if id in self._tasks]
        
        if open_only:
            tasks = [t for t in tasks if t.is_open]
        
        return sorted(tasks, key=lambda t: t.created_at, reverse=True)
    
    def get_pending(self) -> list[Task]:
        """Get all pending (unassigned) tasks."""
        return [t for t in self._tasks.values() if t.status == TaskStatus.PENDING]
    
    def get_blocked(self) -> list[Task]:
        """Get all blocked tasks."""
        return [t for t in self._tasks.values() if t.is_blocked]
    
    def get_overdue(self) -> list[Task]:
        """Get all overdue tasks."""
        return [t for t in self._tasks.values() if t.is_overdue]
    
    def get_by_status(self, status: TaskStatus) -> list[Task]:
        """Get all tasks with a specific status."""
        return [t for t in self._tasks.values() if t.status == status]
    
    def assign(self, task_id: str, agent_id: str) -> Optional[Task]:
        """Assign a task to an agent."""
        task = self.get(task_id)
        if not task:
            return None
        
        # Remove from old assignee index
        if task.assigned_to and task.assigned_to in self._by_assignee:
            if task_id in self._by_assignee[task.assigned_to]:
                self._by_assignee[task.assigned_to].remove(task_id)
        
        # Add to new assignee index
        if agent_id not in self._by_assignee:
            self._by_assignee[agent_id] = []
        self._by_assignee[agent_id].append(task_id)
        
        task.assign(agent_id)
        return task
    
    def list_all(
        self,
        open_only: bool = False,
        limit: int = 100,
    ) -> list[Task]:
        """List all tasks."""
        tasks = list(self._tasks.values())
        
        if open_only:
            tasks = [t for t in tasks if t.is_open]
        
        tasks.sort(key=lambda t: t.created_at, reverse=True)
        return tasks[:limit]
    
    def get_stats(self) -> dict:
        """Get task statistics."""
        all_tasks = list(self._tasks.values())
        
        return {
            "total": len(all_tasks),
            "open": len([t for t in all_tasks if t.is_open]),
            "completed": len([t for t in all_tasks if t.status == TaskStatus.COMPLETED]),
            "failed": len([t for t in all_tasks if t.status == TaskStatus.FAILED]),
            "blocked": len([t for t in all_tasks if t.is_blocked]),
            "overdue": len([t for t in all_tasks if t.is_overdue]),
            "by_priority": {
                p.value: len([t for t in all_tasks if t.priority == p])
                for p in TaskPriority
            },
        }

File: src/coordination/task.ts
/**
 * Task Primitive
 *
 * Tasks are units of work that agents can assign to each other.
 * They provide structure beyond simple messages - tracking state,
 * ownership, deadlines, and completion.
 */

import { z } from 'zod';
import { nanoid } from 'nanoid';

/**
 * Status of a task.
 */
export const TaskStatus = {
  PENDING: 'pending',
  ASSIGNED: 'assigned',
  IN_PROGRESS: 'in_progress',
  BLOCKED: 'blocked',
  IN_REVIEW: 'in_review',
  COMPLETED: 'completed',
  CANCELLED: 'cancelled',
  FAILED: 'failed',
} as const;

export type TaskStatus = (typeof TaskStatus)[keyof typeof TaskStatus];

/**
 * Priority levels for tasks.
 */
export const TaskPriority = {
  CRITICAL: 'critical',
  HIGH: 'high',
  MEDIUM: 'medium',
  LOW: 'low',
  BACKGROUND: 'background',
} as const;

export type TaskPriority = (typeof TaskPriority)[keyof typeof TaskPriority];

export const TaskSchema = z.object({
  id: z.string(),
  title: z.string(),
  description: z.string().default(''),
  createdBy: z.string(),
  assignedTo: z.string().optional(),
  reviewers: z.array(z.string()).default([]),
  status: z.enum(['pending', 'assigned', 'in_progress', 'blocked', 'in_review', 'completed', 'cancelled', 'failed']),
  priority: z.enum(['critical', 'high', 'medium', 'low', 'background']),
  channelId: z.string().optional(),
  threadId: z.string().optional(),
  sourceMessageId: z.string().optional(),
  createdAt: z.date(),
  startedAt: z.date().optional(),
  completedAt: z.date().optional(),
  dueAt: z.date().optional(),
  progressPercent: z.number().min(0).max(100).default(0),
  statusMessage: z.string().default(''),
  inputData: z.record(z.unknown()).default({}),
  outputData: z.record(z.unknown()).default({}),
  blockedBy: z.array(z.string()).default([]),
  blocks: z.array(z.string()).default([]),
  tags: z.array(z.string()).default([]),
});

export type TaskData = z.infer<typeof TaskSchema>;

/**
 * A comment or update on a task.
 */
export interface TaskComment {
  id: string;
  agentId: string;
  content: string;
  createdAt: Date;
  messageId?: string;
}

/**
 * A task assigned to an agent.
 */
export class Task {
  readonly id: string;
  title: string;
  description: string;

  // Ownership
  readonly createdBy: string;
  assignedTo?: string;
  reviewers: string[];

  // Status
  status: TaskStatus;
  priority: TaskPriority;

  // Context links
  channelId?: string;
  threadId?: string;
  sourceMessageId?: string;

  // Timing
  readonly createdAt: Date;
  startedAt?: Date;
  completedAt?: Date;
  dueAt?: Date;

  // Progress
  progressPercent: number;
  statusMessage: string;

  // Input/Output
  inputData: Record<string, unknown>;
  outputData: Record<string, unknown>;

  // Dependencies
  blockedBy: string[];
  blocks: string[];

  // History
  readonly comments: TaskComment[];

  // Metadata
  tags: string[];
  metadata: Record<string, unknown>;

  constructor(data: Partial<TaskData> & { title: string; createdBy: string }) {
    this.id = data.id ?? nanoid();
    this.title = data.title;
    this.description = data.description ?? '';
    this.createdBy = data.createdBy;
    this.assignedTo = data.assignedTo;
    this.reviewers = data.reviewers ?? [];
    this.status = data.status ?? TaskStatus.PENDING;
    this.priority = data.priority ?? TaskPriority.MEDIUM;
    this.channelId = data.channelId;
    this.threadId = data.threadId;
    this.sourceMessageId = data.sourceMessageId;
    this.createdAt = data.createdAt ?? new Date();
    this.startedAt = data.startedAt;
    this.completedAt = data.completedAt;
    this.dueAt = data.dueAt;
    this.progressPercent = data.progressPercent ?? 0;
    this.statusMessage = data.statusMessage ?? '';
    this.inputData = data.inputData ?? {};
    this.outputData = data.outputData ?? {};
    this.blockedBy = data.blockedBy ?? [];
    this.blocks = data.blocks ?? [];
    this.comments = [];
    this.tags = data.tags ?? [];
    this.metadata = {};
  }

  get isOpen(): boolean {
    return (
      this.status !== TaskStatus.COMPLETED &&
      this.status !== TaskStatus.CANCELLED &&
      this.status !== TaskStatus.FAILED
    );
  }

  get isBlocked(): boolean {
    return this.status === TaskStatus.BLOCKED || this.blockedBy.length > 0;
  }

  get isOverdue(): boolean {
    if (!this.dueAt) return false;
    return new Date() > this.dueAt && this.isOpen;
  }

  get duration(): number | undefined {
    if (this.startedAt && this.completedAt) {
      return this.completedAt.getTime() - this.startedAt.getTime();
    }
    return undefined;
  }

  assign(agentId: string): this {
    this.assignedTo = agentId;
    this.status = TaskStatus.ASSIGNED;
    return this;
  }

  start(): this {
    this.status = TaskStatus.IN_PROGRESS;
    this.startedAt = new Date();
    return this;
  }

  updateProgress(percent: number, message = ''): this {
    this.progressPercent = Math.min(100, Math.max(0, percent));
    if (message) {
      this.statusMessage = message;
    }
    return this;
  }

  block(reason: string, blockedByTaskId?: string): this {
    this.status = TaskStatus.BLOCKED;
    this.statusMessage = reason;
    if (blockedByTaskId) {
      this.blockedBy.push(blockedByTaskId);
    }
    return this;
  }

  unblock(taskId?: string): this {
    if (taskId) {
      this.blockedBy = this.blockedBy.filter((id) => id !== taskId);
    }
    if (this.blockedBy.length === 0) {
      this.status = TaskStatus.IN_PROGRESS;
      this.statusMessage = '';
    }
    return this;
  }

  submitForReview(output?: Record<string, unknown>): this {
    this.status = TaskStatus.IN_REVIEW;
    if (output) {
      this.outputData = output;
    }
    this.progressPercent = 100;
    return this;
  }

  complete(output?: Record<string, unknown>): this {
    this.status = TaskStatus.COMPLETED;
    this.completedAt = new Date();
    this.progressPercent = 100;
    if (output) {
      this.outputData = output;
    }
    return this;
  }

  cancel(reason = ''): this {
    this.status = TaskStatus.CANCELLED;
    this.completedAt = new Date();
    this.statusMessage = reason;
    return this;
  }

  fail(reason = '', error?: unknown): this {
    this.status = TaskStatus.FAILED;
    this.completedAt = new Date();
    this.statusMessage = reason;
    if (error) {
      this.outputData.error = String(error);
    }
    return this;
  }

  addComment(agentId: string, content: string, messageId?: string): TaskComment {
    const comment: TaskComment = {
      id: nanoid(),
      agentId,
      content,
      createdAt: new Date(),
      messageId,
    };
    this.comments.push(comment);
    return comment;
  }

  toJSON(): TaskData & { isOpen: boolean; isBlocked: boolean; isOverdue: boolean } {
    return {
      id: this.id,
      title: this.title,
      description: this.description,
      createdBy: this.createdBy,
      assignedTo: this.assignedTo,
      reviewers: this.reviewers,
      status: this.status,
      priority: this.priority,
      channelId: this.channelId,
      threadId: this.threadId,
      sourceMessageId: this.sourceMessageId,
      createdAt: this.createdAt,
      startedAt: this.startedAt,
      completedAt: this.completedAt,
      dueAt: this.dueAt,
      progressPercent: this.progressPercent,
      statusMessage: this.statusMessage,
      inputData: this.inputData,
      outputData: this.outputData,
      blockedBy: this.blockedBy,
      blocks: this.blocks,
      tags: this.tags,
      isOpen: this.isOpen,
      isBlocked: this.isBlocked,
      isOverdue: this.isOverdue,
    };
  }

  toString(): string {
    const statusIcons: Record<TaskStatus, string> = {
      pending: '⏸️',
      assigned: '📋',
      in_progress: '🔄',
      blocked: '🚫',
      in_review: '👀',
      completed: '✅',
      cancelled: '❌',
      failed: '💥',
    };
    const icon = statusIcons[this.status];
    const assignee = this.assignedTo ? ` -> ${this.assignedTo.slice(0, 8)}` : '';
    return `${icon} [${this.priority}] ${this.title}${assignee}`;
  }
}

/**
 * Manages tasks within a workspace.
 */
export class TaskManager {
  private readonly _tasks: Map<string, Task> = new Map();
  private readonly _byAssignee: Map<string, string[]> = new Map();
  private readonly _byChannel: Map<string, string[]> = new Map();

  /**
   * Create a new task.
   */
  create(data: {
    title: string;
    createdBy: string;
    description?: string;
    priority?: TaskPriority;
    assignedTo?: string;
    channelId?: string;
    dueAt?: Date;
    inputData?: Record<string, unknown>;
    tags?: string[];
  }): Task {
    const task = new Task({
      ...data,
      priority: data.priority ?? TaskPriority.MEDIUM,
    });

    if (data.assignedTo) {
      task.status = TaskStatus.ASSIGNED;
    }

    return this.add(task);
  }

  /**
   * Add a task to the manager.
   */
  add(task: Task): Task {
    this._tasks.set(task.id, task);

    // Index by assignee
    if (task.assignedTo) {
      if (!this._byAssignee.has(task.assignedTo)) {
        this._byAssignee.set(task.assignedTo, []);
      }
      this._byAssignee.get(task.assignedTo)!.push(task.id);
    }

    // Index by channel
    if (task.channelId) {
      if (!this._byChannel.has(task.channelId)) {
        this._byChannel.set(task.channelId, []);
      }
      this._byChannel.get(task.channelId)!.push(task.id);
    }

    return task;
  }

  /**
   * Get a task by ID.
   */
  get(taskId: string): Task | undefined {
    return this._tasks.get(taskId);
  }

  /**
   * Get tasks assigned to an agent.
   */
  getByAssignee(agentId: string, options: { openOnly?: boolean } = {}): Task[] {
    const taskIds = this._byAssignee.get(agentId) ?? [];
    let tasks = taskIds.map((id) => this._tasks.get(id)!).filter(Boolean);

    if (options.openOnly !== false) {
      tasks = tasks.filter((t) => t.isOpen);
    }

    return tasks.sort((a, b) => {
      const priorityOrder = { critical: 0, high: 1, medium: 2, low: 3, background: 4 };
      return priorityOrder[a.priority] - priorityOrder[b.priority];
    });
  }

  /**
   * Get tasks in a channel.
   */
  getByChannel(channelId: string, options: { openOnly?: boolean } = {}): Task[] {
    const taskIds = this._byChannel.get(channelId) ?? [];
    let tasks = taskIds.map((id) => this._tasks.get(id)!).filter(Boolean);

    if (options.openOnly !== false) {
      tasks = tasks.filter((t) => t.isOpen);
    }

    return tasks.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
  }

  /**
   * Get all pending (unassigned) tasks.
   */
  getPending(): Task[] {
    return Array.from(this._tasks.values()).filter((t) => t.status === TaskStatus.PENDING);
  }

  /**
   * Get all blocked tasks.
   */
  getBlocked(): Task[] {
    return Array.from(this._tasks.values()).filter((t) => t.isBlocked);
  }

  /**
   * Get all overdue tasks.
   */
  getOverdue(): Task[] {
    return Array.from(this._tasks.values()).filter((t) => t.isOverdue);
  }

  /**
   * Get all tasks with a specific status.
   */
  getByStatus(status: TaskStatus): Task[] {
    return Array.from(this._tasks.values()).filter((t) => t.status === status);
  }

  /**
   * Assign a task to an agent.
   */
  assign(taskId: string, agentId: string): Task | undefined {
    const task = this.get(taskId);
    if (!task) return undefined;

    // Remove from old assignee index
    if (task.assignedTo && this._byAssignee.has(task.assignedTo)) {
      const tasks = this._byAssignee.get(task.assignedTo)!;
      const idx = tasks.indexOf(taskId);
      if (idx !== -1) tasks.splice(idx, 1);
    }

    // Add to new assignee index
    if (!this._byAssignee.has(agentId)) {
      this._byAssignee.set(agentId, []);
    }
    this._byAssignee.get(agentId)!.push(taskId);

    task.assign(agentId);
    return task;
  }

  /**
   * List all tasks.
   */
  listAll(options: { openOnly?: boolean; limit?: number } = {}): Task[] {
    const { openOnly = false, limit = 100 } = options;

    let tasks = Array.from(this._tasks.values());

    if (openOnly) {
      tasks = tasks.filter((t) => t.isOpen);
    }

    tasks.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
    return tasks.slice(0, limit);
  }

  /**
   * Get task statistics.
   */
  getStats(): {
    total: number;
    open: number;
    completed: number;
    failed: number;
    blocked: number;
    overdue: number;
    byPriority: Record<TaskPriority, number>;
  } {
    const allTasks = Array.from(this._tasks.values());

    return {
      total: allTasks.length,
      open: allTasks.filter((t) => t.isOpen).length,
      completed: allTasks.filter((t) => t.status === TaskStatus.COMPLETED).length,
      failed: allTasks.filter((t) => t.status === TaskStatus.FAILED).length,
      blocked: allTasks.filter((t) => t.isBlocked).length,
      overdue: allTasks.filter((t) => t.isOverdue).length,
      byPriority: {
        critical: allTasks.filter((t) => t.priority === TaskPriority.CRITICAL).length,
        high: allTasks.filter((t) => t.priority === TaskPriority.HIGH).length,
        medium: allTasks.filter((t) => t.priority === TaskPriority.MEDIUM).length,
        low: allTasks.filter((t) => t.priority === TaskPriority.LOW).length,
        background: allTasks.filter((t) => t.priority === TaskPriority.BACKGROUND).length,
      },
    };
  }
}
